### DISCUSSION TITLE
Processing EMG data results in null array

### DISCUSSION THREAD

**Discussion by ijhua** [neuropsychology/NeuroKit#469] [Q&A] [answered]
I'm trying to process some EMG data.
It already had some filters (mains + low pass) during recording, so I tried `emg_activation()` first. But, this resulted in an array filled with nans. 
Then, I tried cleaning with `emg_clean()`. This also resulted in the nan array. 
Then, I tried `emg_process()`, and it resulted in 0s or NaNs in all columns other than EMG_Raw. 

I'm really not sure what I'm doing wrong or why these functions don't work with my data. 

Thanks in advance!

**Comment by zen-juen** [comment]
Hi @ijhua, could you try passing your cleaned signal through `emg_amplitude()` first before passing it through `emg_activation()`? The latter function would require the amplitude (i.e., the linear envelope of the emg signal) to be calculated first before it can detect the activation related features. Let us know if this works!

So perhaps something like:
```
emg_cleaned = nk.emg_clean(emg, sampling_rate=1000)
emg_amplitude = nk.emg_amplitude(emg_cleaned)
activity, info = nk.emg_activation(emg_amplitude=emg_amplitude, sampling_rate=1000, method="threshold")
```

**Reply by ijhua** [reply]
That also gives me a null array no matter if I clean the data first or not.

**Reply by zen-juen** [reply]
I see, perhaps you could share with us your script and if possible the emg data (or a sample of it) that you're using so that we can replicate the error and investigate

**Reply by ijhua** [reply]
Sure. Here is a part of my script. I'm trying to upload my data. 

```python
import numpy as np
import neurokit2 as nk
```


```python
gastroc = np.loadtxt("gastroc.txt")
```


```python
gastroc
```




    array([-0.00175  ,  0.002625 ,  0.00525  , ...,  0.0056875,  0.0048125,
           -0.0010625])




```python
emg_cleaned = nk.emg_clean(gastroc)
```


```python
emg_cleaned
```




    array([nan, nan, nan, ..., nan, nan, nan])




```python
emg_amplitude = nk.emg_amplitude(emg_cleaned)
```


```python
emg_amplitude
```




    array([nan, nan, nan, ..., nan, nan, nan])




```python
activity, info = nk.emg_activation(emg_amplitude=emg_amplitude, sampling_rate=10000, method="threshold")
```

    /home/[name]/anaconda3/lib/python3.8/site-packages/neurokit2/events/events_find.py:92: NeuroKitWarning: No events found. Check your event_channel or adjust 'threshold' or 'keep' arguments.
      warn(



```python
activity
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>EMG_Activity</th>
      <th>EMG_Onsets</th>
      <th>EMG_Offsets</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>0.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>0.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>3</th>
      <td>0.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>4</th>
      <td>0.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>11976340</th>
      <td>0.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>11976341</th>
      <td>0.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>11976342</th>
      <td>0.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>11976343</th>
      <td>0.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>11976344</th>
      <td>0.0</td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>11976345 rows × 3 columns</p>
</div>




```python
nk.emg_amplitude(gastroc)
```




    array([nan, nan, nan, ..., nan, nan, nan])




```python

```

**Reply by ijhua** [reply]
[gastroc.zip](https://github.com/neuropsychology/NeuroKit/files/6680820/gastroc.zip)
This is a subset of the original data, but it should give the same results.

**Comment by zen-juen** [comment]
Thanks for sharing this! What's the sampling rate of your data? Do you also happen to have the raw data but without the default filters implemented during recording?

**Reply by ijhua** [reply]
The sampling rate is 10kHz. And sorry no, this is all I have. If I don't apply the filters, then I can't see what the data are like without noise while recording.

**Reply by zen-juen** [reply]
Using the data you provided, I realized there are missing values in the dataset:
```
# Load data
gastroc = np.loadtxt("gastroc.txt")
np.sum(np.isnan(gastroc))
Out[26]: 416
```

For now, I've simply discarded the missing values for convenience but you might be interested in other methods of handling them (e.g., see [tensor-based method here](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=8777071)). After discarding, you can proceed to compute the amplitude (i.e., linear envelope) and extract the activity features. I've skipped the cleaning step since you mentioned that filters were already applied during recording.
```
# Discard missing values
gastroc = gastroc[0:100000]  # Extract 10s of data as example
gastroc = gastroc[~np.isnan(gastroc)]

# Get amplitude
amplitude = nk.emg_amplitude(gastroc)

# Get activity features
activity, info = nk.emg_activation(emg_amplitude=amplitude, sampling_rate=10000, method="threshold",
                                   threshold=0.02, duration_min=0.01)
fig = nk.events_plot([info["EMG_Offsets"], info["EMG_Onsets"]], amplitude)
```
You should get something like this below. Though, bear in mind I've changed the defaults in `emg_activation()` by lowering the threshold for activity detection as well as the minimum duration of the activity as it seems like periods of activity in this signal are quite rapid.
```
info
Out[61]: 
{'EMG_Onsets': array([31032, 76630, 78706, 90090, 92845]),
 'EMG_Offsets': array([31159, 76715, 78790, 90135, 92956]),
 'EMG_Activity': array([31032., 31033., 31034., 31035., 31036., 31037., 31038., 31039.,
        31040., 31041., 31042., 31043., 31044., 31045., 31046., 31047.,
        31048., 31049., 31050., 31051., 31052., 31053., 31054., 31055.,
        31056., 31057., 31058., 31059., 31060., 31061., 31062., 31063.,
        31064., 31065., 31066., 31067., 31068., 31069., 31070., 31071.,
        31072., 31073., 31074., 31075., 31076., 31077., 31078., 31079.,
        31080., 31081., 31082., 31083., 31084., 31085., 31086., 31087.,
        31088., 31089., 31090., 31091., 31092., 31093., 31094., 31095.,
        31096., 31097., 31098., 31099., 31100., 31101., 31102., 31103.,
        31104., 31105., 31106., 31107., 31108., 31109., 31110., 31111.,
        31112., 31113., 31114., 31115., 31116., 31117., 31118., 31119.,
        31120., 31121., 31122., 31123., 31124., 31125., 31126., 31127.,
        31128., 31129., 31130., 31131., 31132., 31133., 31134., 31135.,
        31136., 31137., 31138., 31139., 31140., 31141., 31142., 31143.,
        31144., 31145., 31146., 31147., 31148., 31149., 31150., 31151.,
        31152., 31153., 31154., 31155., 31156., 31157., 31158., 76630.,
        76631., 76632., 76633., 76634., 76635., 76636., 76637., 76638.,
        76639., 76640., 76641., 76642., 76643., 76644., 76645., 76646.,
        76647., 76648., 76649., 76650., 76651., 76652., 76653., 76654.,
        76655., 76656., 76657., 76658., 76659., 76660., 76661., 76662.,
        76663., 76664., 76665., 76666., 76667., 76668., 76669., 76670.,
        76671., 76672., 76673., 76674., 76675., 76676., 76677., 76678.,
        76679., 76680., 76681., 76682., 76683., 76684., 76685., 76686.,
        76687., 76688., 76689., 76690., 76691., 76692., 76693., 76694.,
        76695., 76696., 76697., 76698., 76699., 76700., 76701., 76702.,
        76703., 76704., 76705., 76706., 76707., 76708., 76709., 76710.,
        76711., 76712., 76713., 76714., 78706., 78707., 78708., 78709.,
        78710., 78711., 78712., 78713., 78714., 78715., 78716., 78717.,
        78718., 78719., 78720., 78721., 78722., 78723., 78724., 78725.,
        78726., 78727., 78728., 78729., 78730., 78731., 78732., 78733.,
        78734., 78735., 78736., 78737., 78738., 78739., 78740., 78741.,
        78742., 78743., 78744., 78745., 78746., 78747., 78748., 78749.,
        78750., 78751., 78752., 78753., 78754., 78755., 78756., 78757.,
        78758., 78759., 78760., 78761., 78762., 78763., 78764., 78765.,
        78766., 78767., 78768., 78769., 78770., 78771., 78772., 78773.,
        78774., 78775., 78776., 78777., 78778., 78779., 78780., 78781.,
        78782., 78783., 78784., 78785., 78786., 78787., 78788., 78789.,
        90090., 90091., 90092., 90093., 90094., 90095., 90096., 90097.,
        90098., 90099., 90100., 90101., 90102., 90103., 90104., 90105.,
        90106., 90107., 90108., 90109., 90110., 90111., 90112., 90113.,
        90114., 90115., 90116., 90117., 90118., 90119., 90120., 90121.,
        90122., 90123., 90124., 90125., 90126., 90127., 90128., 90129.,
        90130., 90131., 90132., 90133., 90134., 92845., 92846., 92847.,
        92848., 92849., 92850., 92851., 92852., 92853., 92854., 92855.,
        92856., 92857., 92858., 92859., 92860., 92861., 92862., 92863.,
        92864., 92865., 92866., 92867., 92868., 92869., 92870., 92871.,
        92872., 92873., 92874., 92875., 92876., 92877., 92878., 92879.,
        92880., 92881., 92882., 92883., 92884., 92885., 92886., 92887.,
        92888., 92889., 92890., 92891., 92892., 92893., 92894., 92895.,
        92896., 92897., 92898., 92899., 92900., 92901., 92902., 92903.,
        92904., 92905., 92906., 92907., 92908., 92909., 92910., 92911.,
        92912., 92913., 92914., 92915., 92916., 92917., 92918., 92919.,
        92920., 92921., 92922., 92923., 92924., 92925., 92926., 92927.,
        92928., 92929., 92930., 92931., 92932., 92933., 92934., 92935.,
        92936., 92937., 92938., 92939., 92940., 92941., 92942., 92943.,
        92944., 92945., 92946., 92947., 92948., 92949., 92950., 92951.,
        92952., 92953., 92954., 92955.])}
```
![Figure_1](https://user-images.githubusercontent.com/54053748/122707858-792ac580-d28d-11eb-9447-58781c5b9515.png)

Let me know if this works!

**Reply by DominiqueMakowski** [reply]
(@zen-juen we probably should, for all signals, run linear interpolation by default on NaNs values upon cleaning)

**Reply by zen-juen** [reply]
@DominiqueMakowski yeah, embed them in the `_*clean()` functions and perhaps throw out a warning when this is done?

**Reply by ijhua** [reply]
@zen-juen Thanks!

I have a couple of followup questions.

1
We recorded simultaneously from another muscle. Is there a way to do some analysis like cross-correlation between the 2 muscles or some other measure of coordination?

I've tried 
`nk.signal_synchrony(gastroc,hamstring,method="hilbert")` 
and 
`nk.signal_synchrony(gastroc,hamstring,method="correlation",window_size=500)` 
but both return arrays filled with 1. 
Visually inspecting, I don't think they aren't entirely in phase or out of phase, so I'm not sure how to interpret those results. Here's an example of the 2 signals:
![image](https://user-images.githubusercontent.com/13755787/122784038-1aab1a80-d280-11eb-95c6-553fb4337dcb.png)



2
In the 10s that you extracted, there is 1 contraction of interest. We have another channel that records a square pulse while a stimulus is being applied. I've created events based on that channel, but when I try to use epochs or emg_eventrelated(), it give a DataError. 

Here's the most recent try:
`signals,info = nk.emg_process(gastroc,sampling_rate=10000)`
`epochs = nk.epochs_create(signals, events, sampling_rate=10000,epochs_start=-5,epochs_end=2)`
`nk.epochs_plot(epochs)`

It still makes a plot, but I'm confused about the error `DataError: No numeric types to aggregate.`
![image](https://user-images.githubusercontent.com/13755787/122781537-e9314f80-d27d-11eb-9df7-1eef4bb4f207.png)

And, it's not clear to me how to extract individual periods of activity

**Reply by zen-juen** [reply]
Regarding your question on cross-correlation, it is indeed weird that you're getting an array full of 1s using `nk.signal_synchrony()`, unless they are exact replicas of the same signal. We should be expecting values very close to 1 (rather than 1) instead. Perhaps you could provide subsets of the `gastroc` and `hamstring` data that you're using so that we can check what's wrong.

The DataError is indicating that the data being merged here is incompatible with each other. It also seems like the EMG onsets and offsets are not properly propagated into the plot (refer to the bottom two empty plots in your figure). Could you print out here what your `events` and `epochs` look like?

**Reply by ijhua** [reply]
I've attached a subset of my data. There should be 3 events in this set. The columns are as follows: Time, Gastroc, Hamstring, Trigger, Comments
[sample_EMG.zip](https://github.com/neuropsychology/NeuroKit/files/6690711/sample_EMG.zip)
Trigger contains the square pulse (starts about 5s before the stimulus, and ends after the stimulus ends). I used this channel to create events, and the comments channel to label the events. Since the jump always before the end of the square pulse in trials that it happens, I defined events based on the end of the pulse. 

The gastroc and hamstring values are not exact replicas, so I don't know why the values are 1. Even if they were replicas, we'd expect the correlation to be 0 because the function calculates phase difference right?

Here are the events and epochs:


```
events = nk.events_find(event_channel = trigger, threshold_keep = "above",discard_first=1,event_conditions=comments)
```

```
Out:
{'onset': array([ 2724338,  3405454,  3969976,  4272486,  4583633,  5003269,
         5396625,  5800708,  6235200,  6687240,  7103214,  7531803,
         7902757,  8331590,  8782189,  9100265, 10203853, 10587904,
        11040867, 11300625, 11557529, 11770170, 11952966]),
 'duration': array([572619, 383065, 208802, 215585, 319194, 294444, 291354, 307359,
        248671, 290804, 311579, 266996, 314603, 335292, 213632, 999989,
        282575, 302451, 160014, 160685, 122703,  97238,  23379]),
 'label': array(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12',
        '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23'],
       dtype='<U21'),
 'condition': array(['no jump ', 'jump ', 'jump ', 'jump ', 'jump ', 'jump ', 'jump ',
        'jump ', 'jump - r leg only? ', 'jump ', 'jump ', 'jump ', 'jump ',
        'jump ', 'jump ', 'jump ', 'jump ', 'jump ', 'not good ', 'jump ',
        'jump ', 'jump ', 'jump '], dtype=object)}
```


```
signals,info = nk.emg_process(gastroc,sampling_rate=10000)
epochs = nk.epochs_create(signals, events, sampling_rate=10000,epochs_start=-5,epochs_end=2)
```
The output of ``epochs`` is long, so here's the first entry:

```
Out[]:
{'1':           EMG_Raw  EMG_Clean  EMG_Amplitude  EMG_Activity EMG_Onsets  \
 -5.0000 -0.037937   0.003734       0.000057           0.0          0   
 -4.9999 -0.036312   0.005727       0.000056           0.0          0   
 -4.9998 -0.036187   0.006205       0.000056           0.0          0   
 -4.9997 -0.038375   0.004353       0.000056           0.0          0   
 -4.9996 -0.037812   0.005236       0.000055           0.0          0   
 ...           ...        ...            ...           ...        ...   
  1.9996  0.005125   0.021436       0.000043           0.0          0   
  1.9997  0.006812   0.022485       0.000044           0.0          0   
  1.9998  0.009500   0.024577       0.000044           0.0          0   
  1.9999  0.014063   0.028588       0.000044           0.0          0   
  2.0000  0.018125   0.032145       0.000045           0.0          0   
 
  EMG_Offsets    Index Label Condition  
 -5.0000           0  2674338     1  no jump   
 -4.9999           0  2674339     1  no jump   
 -4.9998           0  2674340     1  no jump   
 -4.9997           0  2674341     1  no jump   
 -4.9996           0  2674342     1  no jump   
 ...             ...      ...   ...       ...  
  1.9996           0  2744333     1  no jump   
  1.9997           0  2744334     1  no jump   
  1.9998           0  2744335     1  no jump   
  1.9999           0  2744336     1  no jump   
  2.0000           0  2744337     1  no jump  
 
 [70000 rows x 9 columns],
```

**Comment by zen-juen** [comment, marked as answer]
I tried running `signal_synchrony` on the data you provided and it seems alright
```
data = pd.read_csv('sample_EMG.csv').iloc[:, 1:6]
data.columns = ['time', 'gastroc', 'hamstring', 'trigger', 'comments']

data[['gastroc', 'hamstring']] = pd.DataFrame.pad(data[['gastroc', 'hamstring']])

nk.signal_synchrony(data['gastroc'], data['hamstring'], method="hilbert")
Out[175]: 
array([0.86847355, 0.62358962, 0.57553423, ..., 0.86722597, 0.85865494,
       0.98653406])

nk.signal_synchrony(data['gastroc'], data['hamstring'], method="correlation", window_size=500)
Out[176]: 
array([-0.47426556, -0.47426556, -0.47426556, ..., -0.47426556,
       -0.47426556, -0.47426556])
```
Could you see if this is what you expected?

---


I managed to replicate the error and realized that DataError is due to the `EMG_Onsets` and `EMG_Offsets` columns being treated as an object rather than float, like so:
```
epochs['1']['EMG_Onsets']
Out[178]: 
-5.0000    NaN
-4.9999    NaN
-4.9998    NaN
-4.9997    NaN
-4.9996    NaN

 1.9996      0
 1.9997      0
 1.9998      0
 1.9999      0
 2.0000      0
Name: EMG_Onsets, Length: 70000, dtype: object
```
And this is problematic for plotting which ideally requires `dtype: float64`. Thanks for bringing this to our attention! I believe that this `dtype: object` is probably the case for all our processed signals which have 0 propagated in them and 1 to indicate some specific feature (e.g., EMG onsets and offsets, ECG peaks...). That being said, we'll probably have to make the necessary changes to the internal functions used in `*_process()` so that this doesn't happen again @DominiqueMakowski 😓 

For now, @ijhua you can run this short code to force the columns to be treated as float and plot them again:
```
for i in epochs:
    epochs[i]['EMG_Onsets'] = pd.to_numeric(epochs[i]['EMG_Onsets'], downcast="float")
    epochs[i]['EMG_Offsets'] = pd.to_numeric(epochs[i]['EMG_Offsets'], downcast="float")

nk.epochs_plot(epochs)
```

![Figure_1](https://user-images.githubusercontent.com/54053748/122861711-7a71f600-d352-11eb-9744-f7849a1f94c7.png)

**Reply by ijhua** [reply]
Ok. I've found out that for `signal_synchrony`, I need to put in the dataframe column, not the values. i.e. Don't enter `df["gastroc"].values` just `df["gastroc"]` Then, I got arrays like you did. 

And thanks for the explanation about the DataError!
